core/distributed/dd_matrix.cpp

```diff
@@ -0,0 +1,416 @@
+void DdMatrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
+    const device_matrix_data<value_type, global_index_type>& data,
+    std::shared_ptr<const Partition<local_index_type, global_index_type>>
+        partition)
+{
+    const auto comm = this->get_communicator();
+    GKO_ASSERT_EQ(data.get_size()[0], partition->get_size());
+    GKO_ASSERT_EQ(data.get_size()[1], partition->get_size());
+    GKO_ASSERT_EQ(comm.size(), partition->get_num_parts());
+    auto exec = this->get_executor();
+    auto local_part = comm.rank();
+    auto use_host_buffer = mpi::requires_host_buffer(exec, comm);
+    auto tmp_partition = make_temporary_clone(exec, partition);
+
+    // set up LinOp sizes
+    auto global_num_rows = partition->get_size();
+    dim<2> global_dim{global_num_rows, global_num_rows};
+    this->set_size(global_dim);
+
+    size_type num_parts = comm.size();
+    array<GlobalIndexType> non_owning_row_idxs{exec};
+    array<GlobalIndexType> non_owning_col_idxs{exec};
+    device_matrix_data<value_type, global_index_type> data_copy{exec, data};
+    auto arrays = data_copy.empty_out();
+
+    exec->run(dd_matrix::make_filter_non_owning_idxs(
+        data, make_temporary_clone(exec, partition).get(),
+        make_temporary_clone(exec, partition).get(), local_part,
+        non_owning_row_idxs, non_owning_col_idxs));
+
+    auto map = gko::experimental::distributed::index_map<LocalIndexType,
+                                                         GlobalIndexType>(
+        exec, partition, local_part, non_owning_row_idxs);
+
+    GlobalIndexType local_num_rows =
+        map.get_local_size() + map.get_non_local_size();
+    auto local_col_idxs = map.map_to_local(
+        arrays.col_idxs, gko::experimental::distributed::index_space::combined);
+    auto local_row_idxs = map.map_to_local(
+        arrays.row_idxs, gko::experimental::distributed::index_space::combined);
+
+    // Construct the local diagonal block.
+    device_matrix_data<value_type, local_index_type> local_data{
+        exec,
+        dim<2>{static_cast<size_type>(local_num_rows),
+               static_cast<size_type>(local_num_rows)},
+        local_row_idxs, local_col_idxs, arrays.values};
+    local_data.sort_row_major();
+    as<ReadableFromMatrixData<ValueType, LocalIndexType>>(this->local_mtx_)
+        ->read(std::move(local_data));
+
+    // Gather local sizes from all ranks and build the partition in the enriched
+    // space.
+    array<GlobalIndexType> range_bounds{
+        use_host_buffer ? exec->get_master() : exec, num_parts + 1};
+    comm.all_gather(exec, &local_num_rows, 1, range_bounds.get_data(), 1);
+    range_bounds.set_executor(exec);
+    exec->run(dd_matrix::make_prefix_sum_nonnegative(range_bounds.get_data(),
+                                                     num_parts + 1));
+    auto large_partition =
+        share(Partition<LocalIndexType, GlobalIndexType>::build_from_contiguous(
+            exec, range_bounds));
```

Comment:
I think all of this is just
```suggestion
    auto large_partition =
        share(build_partition_from_local_size<LocalIndexType, GlobalIndexType>(
            exec, comm, local_num_rows));
```